precompiles/call-permit/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+const PERMIT_DOMAIN: [u8; 32] = keccak256!(
+	"EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+);
+
+#[precompile_utils::generate_function_selector]
+#[derive(Debug, PartialEq)]
+pub enum Action {
+	Dispatch = "dispatch(address,address,uint256,bytes,uint64,uint256,uint8,bytes32,bytes32)",
+	Nonces = "nonces(address)",
+	DomainSeparator = "DOMAIN_SEPARATOR()",
+}
+
+/// Precompile allowing to issue and dispatch call permits for gasless transactions.
+/// A user can sign a permit for a call that can be dispatched and paid by another user or
+/// smart contract.
+pub struct CallPermitPrecompile<Runtime>(PhantomData<Runtime>);
+
+impl<Runtime> Precompile for CallPermitPrecompile<Runtime>
+where
+	Runtime: pallet_evm::Config + pallet_timestamp::Config,
+	<Runtime as pallet_timestamp::Config>::Moment: Into<U256>,
+{
+	fn execute(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		let selector = handle.read_selector()?;
+
+		handle.check_function_modifier(match selector {
+			Action::Dispatch => FunctionModifier::NonPayable,
+			_ => FunctionModifier::View,
+		})?;
+
+		match selector {
+			Action::Dispatch => Self::dispatch(handle),
+			Action::Nonces => Self::nonces(handle),
+			Action::DomainSeparator => Self::domain_separator(handle),
+		}
+	}
+}
+
+impl<Runtime> CallPermitPrecompile<Runtime>
+where
+	Runtime: pallet_evm::Config + pallet_timestamp::Config,
+	<Runtime as pallet_timestamp::Config>::Moment: Into<U256>,
+{
+	fn compute_domain_separator(address: H160) -> [u8; 32] {
+		let name: H256 = keccak_256(b"Call Permit Precompile").into();
+		let version: H256 = keccak256!("1").into();
+		let chain_id: U256 = Runtime::ChainId::get().into();
+
+		let domain_separator_inner = EvmDataWriter::new()
+			.write(H256::from(PERMIT_DOMAIN))
+			.write(name)
+			.write(version)
+			.write(chain_id)
+			.write(Address(address))
+			.build();
+
+		keccak_256(&domain_separator_inner).into()
+	}
+
+	pub fn generate_permit(
+		address: H160,
+		from: H160,
+		to: H160,
+		value: U256,
+		data: Vec<u8>,
+		gaslimit: u64,
+		nonce: U256,
+		deadline: U256,
+	) -> [u8; 32] {
+		let domain_separator = Self::compute_domain_separator(address);
+
+		let permit_content = EvmDataWriter::new()
+			.write(H256::from(PERMIT_TYPEHASH))
+			.write(Address(from))
+			.write(Address(to))
+			.write(value)
+			// bytes are encoded as the keccak_256 of the content
+			.write(H256::from(keccak_256(&data)))
+			.write(gaslimit)
+			.write(nonce)
+			.write(deadline)
+			.build();
+		let permit_content = keccak_256(&permit_content);
+
+		let mut pre_digest = Vec::with_capacity(2 + 32 + 32);
+		pre_digest.extend_from_slice(b"\x19\x01");
+		pre_digest.extend_from_slice(&domain_separator);
+		pre_digest.extend_from_slice(&permit_content);
+		keccak_256(&pre_digest)
+	}
+
+	fn dispatch(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
```

Comment:
Does this cost accurately record the cost in this function? I see at least 2 db reads (chainId, nonces) and one write (nonces). Not to mention the recover_signer, although I agree this one is hardly accoutable yet